multiuser_service_windows.go

```diff
@@ -0,0 +1,295 @@
+				changes <- svc.Status{State: svc.StopPending}
+				log.Printf("Shutting down, received %v", c)
+				interruptChan <- os.Interrupt
+				break loop
+			default:
+				log.Printf("Unexpected control request #%d", c)
+			}
+		}
+	}
+	changes <- svc.Status{State: svc.Stopped}
+	exitCode = <-exitChan
+	return true, exitCode
+}
+
+func runService(name string, isDebug bool) ExitCode {
+	var err error
+	if name == "" {
+		name = "Generic Worker"
+	}
+
+	var elog debug.Log
+	if isDebug {
+		log.Printf("Debug mode enabled, not using eventlog.")
+		elog = debug.New(name)
+	} else {
+		elog, err = eventlog.Open(name)
+		if err != nil || elog == nil {
+			writeCrashFile(CANT_LOG_PROPERLY, err, "Could not open eventlog %q", name)
+			return CANT_LOG_PROPERLY
+		}
+	}
+	logWriter = io.MultiWriter(logWriter, elogWrapper{elog})
+	log.SetOutput(logWriter)
+	// multiwriter will fail if write to any writer fails
+	// so we aggressively handle that scenario
+	err = log.Output(2, fmt.Sprintf("Wrote to eventlog %q successfully", name))
+	if err != nil {
+		writeCrashFile(CANT_LOG_PROPERLY, err, "Unable to log to eventlog %q with writer: %v", name, logWriter)
+		return CANT_LOG_PROPERLY
+	}
+	defer elog.Close()
+
+	log.Printf("Starting service %q", name)
+	run := svc.Run
+	if isDebug {
+		run = debug.Run
+	}
+	err = run(name, &windowsService{})
+	if err != nil {
+		writeCrashFile(INTERNAL_ERROR, err, "Failed to start service %q", name)
+		return CANT_LOG_PROPERLY
+	}
+	// use Output to use all configured loggers and handle err
+	// io.MultiWriter fails for _all_ writers if any fail
+	// this helps us catch that, as normal log.Print* calls
+	// silently eat errors
+	err = log.Output(2, fmt.Sprintf("Stopped service %q", name))
+	if err != nil {
+		writeCrashFile(CANT_LOG_PROPERLY, err, "Unable to log to one or more log outputs, configured writer: %v", logWriter)
+		return CANT_LOG_PROPERLY
+	}
+	return 0
+}
+
+// deploys the generic worker as a windows service named name
+// running as the user LocalSystem
+// if the service already exists we skip.
+func deployService(configFile, name, exePath string, configureForAWS bool, configureForGCP bool) error {
+	m, err := mgr.Connect()
+	if err != nil {
+		return err
+	}
+	defer m.Disconnect()
+	s, err := m.OpenService(name)
+	if err == nil {
+		s.Close()
+		return fmt.Errorf("Service %s already exists", name)
+	}
+	args := []string{
+		"run-service",
+	}
+	if configFile != "" {
+		args = append(args, "--config", configFile)
+	}
+	if configureForAWS {
+		args = append(args, "--configure-for-aws")
+	}
+	if configureForGCP {
+		args = append(args, "--configure-for-gcp")
+	}
+	err = configureService(name, exePath, args)
+	if err != nil {
+		return err
+	}
+	log.Printf("Successfully configured service %q.", name)
+	err = configureEventlogSource(name, exePath)
+	if err != nil {
+		return err
+	}
+	log.Printf("Successfully configured eventlog source %q.", name)
+
+	// start service manually in order to fail fast
```

Comment:
When we create a machine image with generic-worker on it, we typically want to install it without starting it up (since it is just the machine image that real workers will use, and is not a real worker itself), so I think we don't want to start it up here, otherwise it would run on the machine which is not a worker, but it used to create the machine image for the actual workers.